### PR TITLE
Fix some patterns:

### PR DESCRIPTION
**Pull request by julienw**
* Remind <recipient> to <action> on <month> <day> at <hour>
* Remind <recipient> to <action> on the <day>th

### CODE REVIEW

**Comment by julienw** [comment]
@gmarty will you please have a look and tell me how you find this ?

File: src/parsers/chrono/chrono.js [hunk 1]
```diff
@@ -56,6 +56,31 @@ dayPeriodsParser.extract = (text, ref, match) => {
   });
 };
 
+const currentMonthDayParser = Object.assign(
+  new chrono.Parser(),
+  {
+    pattern: () => /\b(?:on|by)\s+the\s+(\d\d?)(?:th|st|rd|nd)\b/,
```
Comment (gmarty) [inline src/parsers/chrono/chrono.js]: - For the sake of consistency, let's reuse what is in the original parsers. See https://github.com/wanasit/chrono/blob/master/src/parsers/EN/ENMonthNameMiddleEndianParser.js#L30 for how the days regexp is usually formatted:

``` js
([0-9]{1,2})(?:st|nd|rd|th)?
```
- Also, note that the suffix is made optional which I think may sense so that 'the 12' is parsed as well.
- Then, all inputs are normalised, so you can safely replace `\s+` by `\s` or even better by just ` ` for greater readibility.
- Finally I wonder if we'll ever get fully written numbers like 'first' or 'second'. That, of course, depends on the output of our speech recogniser. If this is the case, we should probably add a numeric normaliser during the cleaning phase to handle this. The library does through https://github.com/wanasit/chrono/blob/master/src/utils/EN.js#L74-L106 and we can probably import that object as is.

File: src/parsers/action-parser.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ export default class ActionParser {
   parse(obj = { cleaned: '', time: [] }) {
     let action = null;
 
-    if (obj.time === null) {
+    if (!obj.time || !obj.time.length) {
```
Comment (julienw) [inline src/parsers/action-parser.js:42]: This is quite unrelated to the global change, but I found this was quite inconsistent with the default value upthere. Not sure there is an actual bug because of this but I think this makes no harm.

File: src/parsers/chrono/chrono.js [hunk 3]
```diff
@@ -6,8 +6,8 @@ import chrono from 'chrono-node';
  */
 const dayPeriodsParser = new chrono.Parser();
 dayPeriodsParser.pattern =
-  () => new RegExp('midnight|morning|in the morning|noon|' +
-    'afternoon|in the afternoon|evening|in the evening|night|at night', 'i');
+  () => new RegExp('\\b(?:midnight|morning|in the morning|noon|afternoon|' +
+    'in the afternoon|evening|in the evening|night|at night)\\b', 'i');
```
Comment (julienw) [inline src/parsers/chrono/chrono.js:11]: The `\b` means "edge of a word". I think it's always a good practice to have this in regexps matching words. For example we don't want to match `nightly` (I haven't checked we did before the change though, tell me if you'd like me to check this.)

File: src/parsers/chrono/chrono.js (hunk 1 above)
Comment (julienw) [inline src/parsers/chrono/chrono.js]: Adding "on the" or "by the" at the start of the expression should avoid false positives. I'm especially thinking about actions like "meet at the 4th" (implied "street")

File: src/parsers/chrono/chrono.js [hunk 4]
```diff
@@ -56,6 +56,31 @@ dayPeriodsParser.extract = (text, ref, match) => {
...
+      const currentMonth = ref.getMonth();
+
+      const month = day >= currentDay ? currentMonth : currentMonth + 1;
+
+      return new chrono.ParsedResult({
+        ref,
+        text: match[0],
+        index: match.index,
+        start: {
+          month: month + 1,
```
Comment (julienw) [inline src/parsers/chrono/chrono.js:87]: javascript dates have their months between 0 and 11. While chrono expects a month between 1 and 12. Maybe I should have this as a comment, what do you think ?

File: src/parsers/chrono/chrono.js (hunk 3 above)
Comment (gmarty) [reply src/parsers/chrono/chrono.js:11]: That's a good point. I blindly copied code from the library to get started.

File: src/parsers/time-parser.js [hunk 5]
```diff
@@ -31,8 +28,13 @@ export default class TimeParser {
       const start = date.start ? Number(date.start.date()) : null;
       const end = date.end ? Number(date.end.date()) : null;
       const extractedText = date.text;
-      const processedText = text.substr(0, date.index) +
-        text.substr(date.index + extractedText.length);
+      let beforeText = text.substr(0, date.index).trim();
+      const afterText = text.substr(date.index + extractedText.length).trim();
+      if (/ (?:on|by)$/.test(beforeText)) {
+        beforeText = beforeText.slice(0, -3);
```
Comment (julienw) [inline src/parsers/time-parser.js:36]: not very elegant, but because I'm sure the matched text is 3 characters...

**Comment by julienw** [comment]
reference about adding a parser for chrono: chrono parser: https://github.com/wanasit/chrono#parser

**Comment by julienw** [comment]
@tamarahills if you're interested :)

File: src/parsers/time-parser.js [hunk 6]
```diff
@@ -31,8 +28,13 @@ export default class TimeParser {
       const start = date.start ? Number(date.start.date()) : null;
       const end = date.end ? Number(date.end.date()) : null;
       const extractedText = date.text;
-      const processedText = text.substr(0, date.index) +
-        text.substr(date.index + extractedText.length);
+      let beforeText = text.substr(0, date.index).trim();
+      const afterText = text.substr(date.index + extractedText.length).trim();
+      if (/ (?:on|by)$/.test(beforeText)) {
```
Comment (gmarty) [inline src/parsers/time-parser.js]: That looks really hacky for me. This should really be handled by a custom chrono parser.
If it's only a temporary fix, I'd add a comment.

File: src/parsers/chrono/chrono.js (hunk 1 above)
Comment (gmarty) [reply src/parsers/chrono/chrono.js]: There is only so much we can do with pattern matching and at the same time maintaining readible code.
I don't think anybody would naturally say 'Remind me to go shopping at the 4th.'

**Review by gmarty — COMMENTED** [review]
I left some comments that need to be addressed, but overall it's good.
I'd really love to contribute the most generic time parsers to chrono-node. That'd make the code lighter here and show our gratitude to the open source community.

File: src/parsers/time-parser.js (hunk 6 above)
Comment (julienw) [reply src/parsers/time-parser.js]: Well, we're already in a custom chrono parser here. I thought it's the best place to do it because that's where we remove the text recognized by chrono.

maybe  we can file a bug on chrono to recognize "on/by" before a date as well, and add the bug number here in a comment ? What do you think ?

**Comment by julienw** [comment]
@gmarty this is ready for a second review, I added a separate commit with only the changes

File: src/parsers/chrono/chrono.js [hunk 7]
```diff
@@ -59,9 +60,18 @@ dayPeriodsParser.extract = (text, ref, match) => {
...
+      chronoUtils.ORDINAL_WORDS_PATTERN + ')\\b',
+      'i' // flag
+    ),
     extract: (text, ref, match) => {
-      const day = Number(match[1]);
+      const stringMatch = match[1];
+      const numericalDay = match[2];
+      const day = numericalDay
+        ? Number(numericalDay)
+        : chronoUtils.ORDINAL_WORDS[stringMatch.toLowerCase()];
```
Comment (julienw) [inline src/parsers/chrono/chrono.js:74]: maybe this should be in a cleaner instead ?

If yes, I'd like to do it in a separate patch so that we can let this in

File: src/parsers/time-parser.js [hunk 8]
```diff
@@ -31,8 +28,15 @@ export default class TimeParser {
       const start = date.start ? Number(date.start.date()) : null;
       const end = date.end ? Number(date.end.date()) : null;
       const extractedText = date.text;
-      const processedText = text.substr(0, date.index) +
-        text.substr(date.index + extractedText.length);
+      let beforeText = text.substr(0, date.index).trim();
+      const afterText = text.substr(date.index + extractedText.length).trim();
+
+      // temporary fix for https://github.com/wanasit/chrono/issues/152
+      if (/ (?:on|by)$/.test(beforeText)) {
```
Comment (gmarty) [inline src/parsers/time-parser.js]: nit: For the sake of consistency, the pattern should be `/\b(?:on|by)$/`.

**Comment by gmarty** [comment]
I left a nit, but good to land otherwise.
Do you need anything from me to publish the new version on npm?

File: src/parsers/time-parser.js (hunk 8 above)
Comment (julienw) [reply src/parsers/time-parser.js]: yeah well I made this one on prupose to make sure it was 3 characters ;) but well slice(0, -3) works if the space is missing so I guess I can do that.

**Comment by julienw** [comment]
>  Do you need anything from me to publish the new version on npm?

I think you can add me as a comaintener on npm or something like this. But I see you're doing something with yarn too and I know nothing about it yet, so maybe it's better if you push it yourself.

File: src/parsers/chrono/chrono.js (hunk 7 above)
Comment (gmarty) [reply src/parsers/chrono/chrono.js:74]: Can you open a bug for moving this to its one cleaner?
Or maybe add it as an additional step to `ContractionsCleaner`.

Comment (julienw) [reply src/parsers/chrono/chrono.js:74]: done: https://github.com/project-abigail/intent-parser/issues/4
